lib/ui/src/core/store.js

```diff
@@ -1,33 +1,50 @@
-// simple store implementation
-const createStore = initialState => {
-  let state = initialState;
-  const listeners = [];
+// TODO -- make this TS?
 
-  function notify() {
-    listeners.forEach(l => l());
+import { localStorage, sessionStorage } from 'global';
+
+// NOTE that merge mutates the first argument. I think we are OK to mutate the output of xStorage.getItem
+import merge from '../libs/merge';
+
+export const STORAGE_KEY = '@storybook/ui/store';
+
+function get(storage) {
+  const serialized = storage.getItem(STORAGE_KEY);
+  return serialized ? JSON.parse(serialized) : {};
+}
+
+function set(storage, value) {
+  storage.setItem(STORAGE_KEY, JSON.stringify(value));
+}
+
+function update(storage, patch) {
+  const previous = get(storage);
+  set(storage, merge(previous, patch));
+}
+
+// Our store piggybacks off the internal React state of the Context Provider
+// It has been augmented to persist state to local/sessionStorage
+export default class Store {
+  constructor({ setState, getState }) {
+    this.upstreamSetState = setState;
+    this.upstreamGetState = getState;
   }
 
-  function setState(patch) {
-    if (typeof patch === 'function') {
-      state = { ...state, ...patch(state) };
-    } else {
-      state = { ...state, ...patch };
-    }
-    notify();
+  // The assumption is that this will be called once, to initialize the React state
+  // when the module is instanciated
+  getInitialState() {
+    return merge(get(localStorage), get(sessionStorage));
+  }
+
+  getState() {
+    return this.upstreamGetState();
   }
 
-  return {
-    getState() {
-      return state;
-    },
-    setState,
-    subscribe(listener) {
-      listeners.push(listener);
-      return function unsubscribe(l) {
-        listeners.splice(listeners.indexOf(l), 1);
-      };
-    },
-  };
-};
-
-export default createStore;
+  setState(patch, persistence = 'none') {
+    this.upstreamSetState(patch);
+
+    if (persistence !== 'none') {
+      const storage = persistence === 'session' ? sessionStorage : localStorage;
+      update(storage, patch);
```

Comment:
I think some throttling needs to happen.

Writing to localStorage is an expensive, synchronous, blocking action!
And so is JSON.stringify.
Is our state a serialisable object btw?

We can use the second callback argument setState has:

```js
this.write = data => localStorage.setItem(key, JSON.stringify(data));
this.persist = throttle(() => this.getState(), 1000):
this.upstreamSetState(patch, persist);
```

How often to we need to persist? Maybe max once a second?

I like how this makes persistence just an automatic thing.

Would be cool to also persist stuff to the FS someday.

Reply:
Do we WANT to store/persist the entire global state?

It would mean we'd show the stories in the explorer faster, but they may not be accurate at all anymore.

Reply:
Your setState call on store means users can no longer get a callback of when the setState has been fully applied by react.

Since you're taking the second argument of setState, and using it for the persistence enum.

Reply:
OK, a few things here:

> 1. I think some throttling needs to happen.

You are probably right.

> 2. Do we WANT to store/persist the entire global state?

I don't think so. I was just imagining we would persist things that users would expect (settings permanently, things like knobs or UI choices on the session). Other state like stories don't need to be persisted, that's just making trouble for ourselves.

> 3. Would be cool to also persist stuff to the FS someday.
 
Yeah, also to figure out a way to do it per-storybook (right now if you run two storybooks on the same port they will clobber each other no doubt) and across all storybooks (as the persisted state will be attached to a given port). But we can add that.

> 4. Your setState call on store means users can no longer get a callback of when the setState has been fully applied by react.

Yes, I neglected this. What do you think the best approach is? We could just allow either as second argument (enum or callback) or three arguments if desired. 

Or we could always just have three arguments `setState(patch, persistence, callback)`. 

What do you think?

Reply:
Oh and:

> 5. Is our state a serialisable object btw?

I think it's the responsibility of the caller to ensure that what they are persisting is serializable. If they aren't persisting, they can do what they want of course.